<a href="https://colab.research.google.com/github/usman-infinity/AIC_345/blob/main/MLF_lab_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Introduction to Multiple Linear Regression

### Definition and Purpose
Multiple Linear Regression (MLR) is a statistical technique used to predict the outcome of a dependent variable based on the values of *two or more* independent variables. Its primary objective is to model the linear relationship between these multiple predictor variables and a single response variable.

The core idea is to find a linear equation that best describes how the dependent variable changes as the independent variables change. This equation helps in understanding the magnitude and direction of the influence each independent variable has on the dependent variable.

### Difference from Simple Linear Regression (SLR)

The key distinction between Multiple Linear Regression and Simple Linear Regression lies in the number of independent variables:

*   **Simple Linear Regression (SLR):** Involves one dependent variable and **one independent variable**. The model seeks to establish a linear relationship between these two variables.
    *   *Equation Example:* `Y = b0 + b1*X1 + e`
*   **Multiple Linear Regression (MLR):** Involves one dependent variable and **two or more independent variables**. The model aims to find a linear relationship that incorporates the effects of multiple predictors.
    *   *Equation Example:* `Y = b0 + b1*X1 + b2*X2 + ... + bn*Xn + e`

Where:
*   `Y` is the dependent (response) variable.
*   `X1, X2, ..., Xn` are the independent (predictor) variables.
*   `b0` is the intercept (the value of Y when all X's are zero).
*   `b1, b2, ..., bn` are the coefficients for each independent variable, representing the change in Y for a one-unit change in the respective X, holding other X's constant.
*   `e` is the error term.

### When is Multiple Linear Regression Used?

MLR is particularly useful in scenarios where a phenomenon is influenced by several factors simultaneously. It allows for a more comprehensive understanding and more accurate predictions compared to relying on a single predictor.

Common use cases include:

*   **Predicting house prices:** Factors like size (square footage), number of bedrooms, location, age of the house, and proximity to amenities all influence the price.
*   **Forecasting sales:** Advertising expenditure, economic indicators, seasonal factors, and competitor pricing can all impact sales figures.
*   **Understanding academic performance:** Student grades might be influenced by study hours, attendance, prior test scores, and socio-economic background.
*   **Assessing health outcomes:** A patient's recovery time could depend on age, pre-existing conditions, medication dosage, and lifestyle.

### Advantage of Multiple Predictors

Using multiple predictors in a linear regression model offers significant advantages over a single predictor:

1.  **Improved Accuracy:** By incorporating more relevant information, MLR models can often explain a larger proportion of the variance in the dependent variable, leading to more accurate predictions.
2.  **Enhanced Explanatory Power:** MLR allows researchers to understand the individual contribution and statistical significance of each independent variable while controlling for others. This provides a more nuanced insight into the relationships within the data.
3.  **Better Model Fit:** A model with multiple predictors can often achieve a better fit to the observed data, as it can capture more complex relationships than a simple linear model.

## Mathematical Formulation




```markdown
## Mathematical Formulation: Multiple Linear Regression

The mathematical model for Multiple Linear Regression is expressed as:

$$y = \beta_0 + \beta_1x_1 + \beta_2x_2 + \dots + \beta_nx_n + \epsilon$$

Where:

*   `y`: This is the **dependent variable** (also known as the response variable), which is the outcome we are trying to predict or explain.
*   `x1, x2, ..., xn`: These are the **independent variables** (also known as predictor variables or features). They are the variables used to predict `y`.
*   `β0` (beta naught): This is the **y-intercept**. It represents the expected mean value of `y` when all independent variables (`x1, x2, ..., xn`) are equal to zero.
*   `β1, β2, ..., βn` (beta coefficients): These are the **coefficients** for each independent variable. `βi` represents the expected change in `y` for a one-unit change in `xi`, assuming all other independent variables remain constant.
*   `ε` (epsilon): This is the **error term** (or residual). It accounts for the variability in `y` that cannot be explained by the independent variables included in the model. It captures unobserved factors and random noise.

### Assumptions of Linear Regression:

For the statistical inference of linear regression models to be valid, several assumptions are made about the error term and the relationship between variables:

1.  **Linearity**: The relationship between the independent variables and the dependent variable is linear.
2.  **Independence of Errors**: The errors (residuals) are independent of each other. This means there is no correlation between consecutive errors.
3.  **Homoscedasticity**: The variance of the errors is constant across all levels of the independent variables. In simpler terms, the spread of the residuals should be roughly the same along the regression line.
4.  **Normality of Errors**: The errors are normally distributed. This assumption is particularly important for constructing confidence intervals and performing hypothesis tests.
5.  **No Multicollinearity**: The independent variables are not highly correlated with each other. High multicollinearity can make it difficult to estimate the individual effects of the independent variables.
```

## Real-life Example Setup




In [ ]:
import numpy as np
import pandas as pd

print("Libraries imported successfully.")

Libraries imported successfully.


In [ ]:
np.random.seed(42) # for reproducibility

# Define number of samples
n_samples = 1000

# Generate independent variables
size = np.random.normal(loc=2000, scale=500, size=n_samples) # Size in sqft
bedrooms = np.random.randint(2, 6, size=n_samples) # Number of bedrooms
distance_to_city = np.random.normal(loc=10, scale=3, size=n_samples) # Distance to city center in miles

# Define coefficients and intercept
intercept = 50000
coeff_size = 150 # $/sqft
coeff_bedrooms = 20000 # $/bedroom
coeff_distance = -5000 # $/mile (closer to city, higher price)

# Generate dependent variable (House Price) with noise
noise = np.random.normal(loc=0, scale=25000, size=n_samples) # Add some random noise
house_price = intercept + (coeff_size * size) + (coeff_bedrooms * bedrooms) + (coeff_distance * distance_to_city) + noise

# Ensure house prices are not negative (though unlikely with these parameters)
house_price[house_price < 0] = 0

# Create a Pandas DataFrame
df = pd.DataFrame({
    'Size (sqft)': size,
    'Number of Bedrooms': bedrooms,
    'Distance to City Center (miles)': distance_to_city,
    'House Price ($)': house_price
})

print("Synthetic Dataset Generated:")
print(df.head())

print("\n--- Variable Definitions ---")
print("Dependent Variable (Target): 'House Price ($)'")
print("Independent Variables (Features): 'Size (sqft)', 'Number of Bedrooms', 'Distance to City Center (miles)'")

Synthetic Dataset Generated:
   Size (sqft)  Number of Bedrooms  Distance to City Center (miles)  \
0  2248.357077                   4                         8.904035   
1  1930.867849                   5                        10.554041   
2  2323.844269                   2                         5.958621   
3  2761.514928                   5                         7.085158   
4  1882.923313                   4                        13.601242   

   House Price ($)  
0    420647.435501  
1    350618.842332  
2    385737.039372  
3    503702.515506  
4    349613.971512  

--- Variable Definitions ---
Dependent Variable (Target): 'House Price ($)'
Independent Variables (Features): 'Size (sqft)', 'Number of Bedrooms', 'Distance to City Center (miles)'


## Model Implementation and Training




In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Define independent variables (X) and dependent variable (y)
X = df[['Size (sqft)', 'Number of Bedrooms', 'Distance to City Center (miles)']]
y = df['House Price ($)']

# Split the data into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data split into training and testing sets successfully.")
print(f"X_train shape: {X_train.shape}, y_train shape: {y_train.shape}")
print(f"X_test shape: {X_test.shape}, y_test shape: {y_test.shape}")

# Instantiate the Linear Regression model
model = LinearRegression()

# Train the model using the training data
model.fit(X_train, y_train)

print("\nLinear Regression model trained successfully.")

# Print the intercept and coefficients
print(f"\nModel Intercept: {model.intercept_:.2f}")
print("Model Coefficients:")
for feature, coef in zip(X.columns, model.coef_):
    print(f"  {feature}: {coef:.2f}")

Data split into training and testing sets successfully.
X_train shape: (800, 3), y_train shape: (800,)
X_test shape: (200, 3), y_test shape: (200,)

Linear Regression model trained successfully.

Model Intercept: 43355.98
Model Coefficients:
  Size (sqft): 149.05
  Number of Bedrooms: 20928.71
  Distance to City Center (miles): -4452.38


### Interpretation of Learned Coefficients

The trained Multiple Linear Regression model provides the following intercept and coefficients:

*   **Model Intercept:** `43355.98`
*   **Coefficient for 'Size (sqft)':** `149.05`
*   **Coefficient for 'Number of Bedrooms':** `20928.71`
*   **Coefficient for 'Distance to City Center (miles)':** `-4452.38`

Let's interpret these values in the context of house price prediction:

1.  **Intercept (\$43,355.98):**
    *   The intercept represents the estimated average house price when all independent variables (Size, Number of Bedrooms, Distance to City Center) are zero. In this specific context, a house with 0 square feet, 0 bedrooms, and 0 miles distance to the city center is not realistic. Therefore, the intercept should generally be interpreted with caution and primarily as a mathematical component of the regression equation rather than a practical prediction for a house with zero features. However, it forms the base price from which other factors add or subtract.

2.  **Coefficient for 'Size (sqft)' (\$149.05):**
    *   This coefficient indicates that, holding the number of bedrooms and distance to the city center constant, for every one-square-foot increase in the house's size, the house price is estimated to increase by approximately \$149.05.

3.  **Coefficient for 'Number of Bedrooms' (\$20,928.71):**
    *   This coefficient suggests that, holding house size and distance to the city center constant, for every additional bedroom, the house price is estimated to increase by approximately \$20,928.71.

4.  **Coefficient for 'Distance to City Center (miles)' (-\$4,452.38):**
    *   This coefficient implies that, holding house size and number of bedrooms constant, for every one-mile increase in the distance from the city center, the house price is estimated to decrease by approximately \$4,452.38. Conversely, a house that is one mile *closer* to the city center (i.e., a one-unit decrease in distance) is estimated to have a \$4,452.38 higher price.

## Evaluation Metrics




In [ ]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate R-squared
r2 = r2_score(y_test, y_pred)

# Calculate Mean Absolute Error (MAE)
mae = mean_absolute_error(y_test, y_pred)

# Calculate Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)

# Calculate Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)

print("\n--- Model Evaluation on Test Set ---")
print(f"R-squared: {r2:.4f}")
print(f"Mean Absolute Error (MAE): ${mae:,.2f}")
print(f"Mean Squared Error (MSE): ${mse:,.2f}")
print(f"Root Mean Squared Error (RMSE): ${rmse:,.2f}")


--- Model Evaluation on Test Set ---
R-squared: 0.9057
Mean Absolute Error (MAE): $20,276.81
Mean Squared Error (MSE): $619,733,082.77
Root Mean Squared Error (RMSE): $24,894.44


### Explanation of Evaluation Metrics

After training the Multiple Linear Regression model and making predictions on the test set, we evaluate its performance using several key regression metrics. These metrics help us understand how well our model fits the data and how accurately it can predict new, unseen values.

Here's an explanation of the metrics calculated:

1.  **R-squared (Coefficient of Determination)**
    *   **Definition:** R-squared measures the proportion of the variance in the dependent variable that is predictable from the independent variables. It's a statistical measure that represents how well the regression model explains the observed variability of the response variable.
    *   **Interpretation:** An R-squared value ranges from 0 to 1 (or 0% to 100%).
        *   An R-squared of 1 (or 100%) indicates that the model explains all the variability of the response variable around its mean.
        *   An R-squared of 0 indicates that the model explains none of the variability of the response variable around its mean.
    *   **In our case (R-squared: 0.9057):** This means that approximately 90.57% of the variance in house prices can be explained by the independent variables (size, number of bedrooms, and distance to city center) in our model. This is a relatively high value, suggesting a good fit.

2.  **Mean Absolute Error (MAE)**
    *   **Definition:** MAE is the average of the absolute differences between the predicted and actual values. It measures the average magnitude of the errors in a set of predictions, without considering their direction.
    *   **Interpretation:** MAE is expressed in the same units as the dependent variable. A lower MAE indicates a more accurate model.
    *   **In our case (MAE: \$20,276.81):** On average, our model's predictions for house prices deviate from the actual house prices by about \$20,276.81. This gives a straightforward understanding of the typical prediction error.

3.  **Mean Squared Error (MSE)**
    *   **Definition:** MSE is the average of the squared differences between the predicted and actual values. By squaring the errors, MSE penalizes larger errors more heavily than smaller ones.
    *   **Interpretation:** MSE is expressed in squared units of the dependent variable, which can make it less intuitive to interpret directly than MAE or RMSE. A lower MSE indicates a more accurate model.
    *   **In our case (MSE: \$619,733,082.77):** This value is large because it's in squared dollars. Its primary use is often in mathematical optimization during model training.

4.  **Root Mean Squared Error (RMSE)**
    *   **Definition:** RMSE is the square root of the MSE. It reverts the error units back to the original units of the dependent variable, making it more interpretable than MSE.
    *   **Interpretation:** Like MAE, RMSE is expressed in the same units as the dependent variable. It provides a measure of the typical magnitude of the errors, but because of the squaring operation, it gives higher weight to larger errors. RMSE is often preferred over MAE when large errors are particularly undesirable.
    *   **In our case (RMSE: \$24,894.44):** This means that the standard deviation of the residuals (prediction errors) is approximately \$24,894.44. Similar to MAE, this gives us an idea of the typical error size, but with a stronger penalty for outliers.

## Interpretation of Evaluation Results

Let's review the calculated evaluation metrics for our Multiple Linear Regression model on the house price prediction test set:

*   **R-squared:** `0.9057`
*   **Mean Absolute Error (MAE):** `20,276.81`
*   **Mean Squared Error (MSE):** `619,733,082.77`
*   **Root Mean Squared Error (RMSE):** `24,894.44`

#### 1. R-squared (Coefficient of Determination)

*   **Significance:** R-squared measures the proportion of the variance in the dependent variable (House Price) that can be explained by the independent variables (Size, Number of Bedrooms, Distance to City Center) in our model. It ranges from 0 to 1, where higher values indicate a better fit.
*   **Our Result (0.9057):** An R-squared value of `0.9057` (or approximately 90.57%) is quite good. This means that about 90.57% of the variability in house prices can be explained by our model's chosen features. The remaining 9.43% of the variability is still unexplained by the model, likely due to other factors not included in our dataset or inherent randomness.

#### 2. Mean Absolute Error (MAE)

*   **Significance:** MAE measures the average magnitude of the errors in a set of predictions, without considering their direction. It is the average of the absolute differences between the predicted and actual values. MAE is expressed in the same units as the dependent variable, making it easily interpretable.
*   **Our Result ($20,276.81):** This means that, on average, our model's predictions for house prices are off by approximately **$20,276.81** from the actual prices. This provides a straightforward measure of the typical prediction error.

#### 3. Mean Squared Error (MSE)

*   **Significance:** MSE measures the average of the squares of the errors. By squaring the errors, MSE penalizes larger errors more heavily than smaller ones, making it sensitive to outliers. Like MAE, it is also in the units of the dependent variable, but squared.
*   **Our Result ($619,733,082.77):** Due to the squaring of errors, the value of MSE (approximately $619.7 million) is much larger and less intuitive for direct interpretation in dollar terms. Its primary role is often in the optimization process of many regression algorithms (like Ordinary Least Squares), where minimizing MSE leads to the 'best fit' line, as it seeks to minimize the total squared distance between the predicted and actual values. However, for presenting model performance to a non-technical audience, it's less practical than MAE or RMSE.

#### 4. Root Mean Squared Error (RMSE)

*   **Significance:** RMSE is the square root of MSE. It brings the error metric back into the same units as the dependent variable, making it more interpretable than MSE. Like MSE, it still gives relatively high weight to large errors.
*   **Our Result ($24,894.44):** The RMSE of **$24,894.44** tells us that the standard deviation of the residuals (prediction errors) is approximately this amount. It can be interpreted as the typical distance between the predicted values and the actual observed values. The fact that RMSE ($24,894.44) is slightly higher than MAE ($20,276.81) suggests that there might be some larger individual prediction errors (outliers) that are being more heavily penalized by RMSE.

#### Overall Assessment of Model Performance

Considering these metrics, the Multiple Linear Regression model performs reasonably well in predicting house prices:

*   The **high R-squared (0.9057)** indicates that the selected independent variables are strong predictors, explaining a significant portion of the variance in house prices.
*   The **MAE ($20,276.81)** and **RMSE ($24,894.44)** provide a tangible understanding of the model's prediction accuracy. An average error of around $20,000-$25,000 for house prices, which can range from hundreds of thousands to millions, suggests a decent predictive capability, especially for a simplified synthetic dataset. The difference between MAE and RMSE suggests that while most predictions are close, there might be a few instances where the model makes larger errors.

In a real-world scenario, whether these error margins are acceptable would depend on the specific application and business requirements. However, for a foundational model, these results demonstrate a good grasp of the underlying relationships.

## Practice Question: Predicting Car Prices

**Scenario:**
You are a data scientist at a car dealership and want to predict the price of used cars based on several features. You have collected a dataset with the following information for various used cars:

*   **Mileage (miles):** The total distance the car has traveled.
*   **Age (years):** The age of the car since its manufacturing date.
*   **Engine Size (liters):** The volume of the car's engine.
*   **Brand (categorical):** The brand of the car (e.g., 'Toyota', 'Honda', 'BMW', 'Mercedes').
*   **Price ($):** The selling price of the car (your target variable).

**Instructions:**

1.  **Generate a Synthetic Dataset:**
    *   Create a synthetic dataset (similar to the house price example) for car prices with at least 1000 samples. Include 'Mileage', 'Age', 'Engine Size', and 'Price' as numerical variables. For 'Brand', you can generate a categorical variable with a few distinct brands and assign different price impacts to them.
    *   Ensure there's a linear relationship between the independent variables and the car 'Price', adding some random noise.
    *   Clearly define your dependent and independent variables.

2.  **Implement and Train a Multiple Linear Regression Model:**
    *   Split your synthetic dataset into training and testing sets (e.g., 80% training, 20% testing).
    *   Train a Multiple Linear Regression model using `scikit-learn` on your training data.
    *   *Hint: For the categorical 'Brand' variable, you will need to apply one-hot encoding or a similar technique before training the model.*

3.  **Interpret Coefficients:**
    *   Print the intercept and coefficients of your trained model.
    *   Interpret each coefficient in the context of predicting car prices. Explain what a positive or negative coefficient signifies for each feature.

4.  **Evaluate Model Performance:**
    *   Make predictions on your test set.
    *   Calculate and print the R-squared, Mean Absolute Error (MAE), Mean Squared Error (MSE), and Root Mean Squared Error (RMSE) for your model on the test set.

5.  **Interpret Evaluation Results:**
    *   Explain what each calculated metric (R-squared, MAE, MSE, RMSE) signifies regarding your model's performance and its predictive power for car prices. Discuss whether you think the model is performing well, and why.

## Introduction to Logistic Regression

Logistic Regression is a statistical model used for **binary classification** tasks, meaning it predicts the probability that an observation belongs to one of two classes. Despite its name, Logistic Regression is a classification algorithm, not a regression algorithm in the traditional sense, as its output is a probability that can be mapped to a discrete class.

### Purpose: Binary Classification
The primary purpose of Logistic Regression is to model the probability of a binary outcome. For example, it can predict whether an email is spam (1) or not spam (0), whether a customer will churn (1) or not churn (0), or whether a patient has a disease (1) or not (0).

The model estimates the probability of a given input belonging to a particular class by fitting data to a **logistic function** (also known as the sigmoid function).

### How it Differs from Linear Regression
While both Linear Regression and Logistic Regression are generalized linear models, they differ significantly in their purpose and output:

1.  **Output Type:**
    *   **Linear Regression:** Predicts a continuous numerical value (e.g., house price, temperature). The output can range from negative infinity to positive infinity.
    *   **Logistic Regression:** Predicts the probability of a binary outcome. The output is constrained to be between 0 and 1, which represents the probability of the event occurring.

2.  **Mathematical Function:**
    *   **Linear Regression:** Uses a linear function (e.g., `Y = b0 + b1*X1 + ...`) directly to predict the output.
    *   **Logistic Regression:** Uses a linear function as an input to a **sigmoid (logistic) function**. The sigmoid function transforms the linear output into a probability. The formula is typically expressed as:
        $$P(Y=1|X) = \frac{1}{1 + e^{-(\beta_0 + \beta_1X_1 + \dots + \beta_nX_n)}}$$
        Where $P(Y=1|X)$ is the probability that the dependent variable $Y$ is 1 given the independent variables $X$.

3.  **Dependent Variable:**
    *   **Linear Regression:** Dependent variable is continuous.
    *   **Logistic Regression:** Dependent variable is categorical (binary).

### When is Logistic Regression Used?
Logistic Regression is widely used in various fields for predicting binary outcomes. Some common real-world scenarios include:

*   **Healthcare:** Predicting the likelihood of a patient developing a certain disease (e.g., diabetes, heart disease) based on medical history, symptoms, and test results.
*   **Finance:** Assessing the probability of loan default by a customer based on their credit score, income, and debt-to-income ratio; fraud detection.
*   **Marketing:** Predicting whether a customer will purchase a product or churn from a service based on their demographics, browsing history, and past interactions.
*   **Spam Detection:** Classifying emails as spam or not spam based on keywords, sender, and other email characteristics.
*   **Quality Control:** Determining whether a product is defective (pass/fail) based on manufacturing parameters.


## Mathematical Formulation: Logistic Regression

Logistic Regression is a statistical model used for binary classification tasks. Unlike Linear Regression, which predicts a continuous outcome, Logistic Regression predicts the probability of a binary outcome (e.g., 0 or 1, true or false, yes or no).

### 1. The Logistic (Sigmoid) Function

The core of Logistic Regression lies in the **logistic function**, also known as the **sigmoid function**. This function transforms a linear combination of independent variables into a probability value that lies between 0 and 1.

The linear combination of independent variables is typically represented as:

$$z = \beta_0 + \beta_1x_1 + \beta_2x_2 + \dots + \beta_nx_n$$

Where:
*   `z`: The linear combination of independent variables.
*   `\u03b20`: The intercept.
*   `\u03b21, \u03b22, ..., \u03b2n`: The coefficients for each independent variable.
*   `x1, x2, ..., xn`: The independent variables.

The sigmoid function, denoted as `\u03c3(z)` or `P(y=1|X)`, is defined as:

$$P(y=1|X) = \sigma(z) = \frac{1}{1 + e^{-z}}$$

Where:
*   `P(y=1|X)`: The probability that the dependent variable `y` is 1, given the independent variables `X`.
*   `e`: The base of the natural logarithm (approximately 2.71828).
*   `z`: The linear combination calculated above.

This function squashes any real-valued number `z` into a value between 0 and 1, making it suitable for interpreting as a probability.

### 2. The Concept of Log-Odds

The linear combination `z` in Logistic Regression can be interpreted as the **log-odds** of the event occurring. The odds of an event are defined as the ratio of the probability of the event occurring to the probability of the event not occurring:

$$\text{Odds} = \frac{P(y=1|X)}{P(y=0|X)} = \frac{P(y=1|X)}{1 - P(y=1|X)}$$

Taking the natural logarithm of the odds gives us the log-odds:

$$\text{Log-odds} = \ln\left(\frac{P(y=1|X)}{1 - P(y=1|X)}\right)$$

It can be shown that this log-odds expression is exactly equal to the linear combination `z`:

$$\ln\left(\frac{P(y=1|X)}{1 - P(y=1|X)}\right) = \beta_0 + \beta_1x_1 + \beta_2x_2 + \dots + \beta_nx_n$$

Thus, Logistic Regression models the log-odds of the dependent variable as a linear combination of the independent variables. The coefficients `\u03b2i` represent the change in the log-odds for a one-unit change in the respective `xi`, holding other variables constant.

### 3. How it Outputs Probabilities

The model outputs probabilities by applying the sigmoid function to the linear combination of predictors. Since the output of the sigmoid function always falls between 0 and 1, these values can be directly interpreted as probabilities. For a given set of input features `X`, the model calculates `z` and then transforms it using the sigmoid function to get `P(y=1|X)`.

For example:
*   If `z` is a large positive number, `e^-z` approaches 0, so `P(y=1|X)` approaches 1.
*   If `z` is 0, `e^-z` is 1, so `P(y=1|X)` is 0.5.
*   If `z` is a large negative number, `e^-z` approaches infinity, so `P(y=1|X)` approaches 0.

To make a final classification decision, a threshold (commonly 0.5) is applied to these probabilities. If `P(y=1|X) >= 0.5`, the prediction is 1; otherwise, it's 0.

### Assumptions of Logistic Regression:

While Logistic Regression does not assume linearity between the independent and dependent variables themselves, it does make several key assumptions:

1.  **Binary Outcome**: The dependent variable must be binary (dichotomous).
2.  **Independence of Observations**: Observations should be independent of each other.
3.  **No Multicollinearity**: The independent variables should not be highly correlated with each other. High multicollinearity can make it difficult to estimate the individual effects of the independent variables.
4.  **Linearity of Log-Odds**: There is a linear relationship between the independent variables and the log-odds of the dependent variable. This is the crucial 'linearity' assumption for Logistic Regression.
5.  **Large Sample Size**: Logistic Regression tends to perform better with larger sample sizes.

## Real-life Example Setup




In [ ]:
import numpy as np
import pandas as pd

# 2. Set a random seed for reproducibility
np.random.seed(42)

# 3. Define the number of samples for the dataset
n_samples = 1000

# 4. Generate synthetic data for independent numerical variables
# Monthly Charges (e.g., average bill between $50 and $100)
monthly_charges = np.random.normal(loc=75, scale=15, size=n_samples)
monthly_charges[monthly_charges < 0] = 0 # Ensure non-negative charges

# Tenure in Months (e.g., customers stay between 1 and 60 months)
tenure_months = np.random.randint(1, 61, size=n_samples)

# Support Calls (e.g., 0 to 5 calls per month)
support_calls = np.random.randint(0, 6, size=n_samples)

# 5. Define coefficients and an intercept
# These coefficients influence the log-odds of churn
intercept = -2.0 # Base log-odds of churn
coeff_monthly_charges = 0.05 # Higher charges might increase churn likelihood
coeff_tenure_months = -0.03 # Longer tenure might decrease churn likelihood
coeff_support_calls = 0.4 # More support calls might increase churn likelihood

# 6. Calculate the linear combination 'z'
z = intercept + (coeff_monthly_charges * monthly_charges) + \
    (coeff_tenure_months * tenure_months) + \
    (coeff_support_calls * support_calls)

# 7. Apply the sigmoid function to 'z' to transform it into churn probabilities
churn_probability = 1 / (1 + np.exp(-z))

# 8. Generate the binary dependent variable, 'Churn'
# If a random number is less than the probability, it's a churn (1), otherwise no churn (0)
churn = (np.random.rand(n_samples) < churn_probability).astype(int)

# 9. Create a Pandas DataFrame named 'churn_df'
churn_df = pd.DataFrame({
    'Monthly_Charges': monthly_charges,
    'Tenure_Months': tenure_months,
    'Support_Calls': support_calls,
    'Churn': churn
})

print("Synthetic Dataset for Customer Churn Generated:")
# 10. Display the first few rows of the churn_df DataFrame
print(churn_df.head())

print("\n--- Variable Definitions ---")
# 11. Explicitly state the name of the dependent variable and the names of the independent variables
print("Dependent Variable (Target): 'Churn'")
print("Independent Variables (Features): 'Monthly_Charges', 'Tenure_Months', 'Support_Calls'")

Synthetic Dataset for Customer Churn Generated:
   Monthly_Charges  Tenure_Months  Support_Calls  Churn
0        82.450712             55              3      0
1        72.926035             44              3      0
2        84.715328             49              2      1
3        97.845448             40              1      1
4        71.487699             11              5      0

--- Variable Definitions ---
Dependent Variable (Target): 'Churn'
Independent Variables (Features): 'Monthly_Charges', 'Tenure_Months', 'Support_Calls'


## Model Implementation and Training




In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

# Define independent variables (X) and dependent variable (y)
X = churn_df[['Monthly_Charges', 'Tenure_Months', 'Support_Calls']]
y = churn_df['Churn']

print("X and y defined successfully.")

# Split the data into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data split into training and testing sets successfully.")
print(f"X_train shape: {X_train.shape}, y_train shape: {y_train.shape}")
print(f"X_test shape: {X_test.shape}, y_test shape: {y_test.shape}")

# Instantiate the Logistic Regression model
model = LogisticRegression(random_state=42, solver='liblinear') # Using liblinear for smaller datasets and good for L1/L2 regularization

# Train the model using the training data
model.fit(X_train, y_train)

print("\nLogistic Regression model trained successfully.")

X and y defined successfully.
Data split into training and testing sets successfully.
X_train shape: (800, 3), y_train shape: (800,)
X_test shape: (200, 3), y_test shape: (200,)

Logistic Regression model trained successfully.


## Interpretation of Coefficients



In [ ]:
print("\nModel Intercept:", model.intercept_[0])
print("Model Coefficients:")
for feature, coef in zip(X.columns, model.coef_[0]):
    print(f"  {feature}: {coef:.4f}")

# Optional: Calculate and print Odds Ratios
print("\nOdds Ratios (exp(coefficient)):")
for feature, coef in zip(X.columns, model.coef_[0]):
    print(f"  {feature}: {np.exp(coef):.4f}")


Model Intercept: -1.5969061508042768
Model Coefficients:
  Monthly_Charges: 0.0463
  Tenure_Months: -0.0352
  Support_Calls: 0.3825

Odds Ratios (exp(coefficient)):
  Monthly_Charges: 1.0474
  Tenure_Months: 0.9654
  Support_Calls: 1.4659


### Interpretation of Learned Coefficients

For our trained Logistic Regression model, the intercept and coefficients are as follows:

*   **Model Intercept (Log-Odds):** `-1.5969`
*   **Coefficient for 'Monthly_Charges':** `0.0463`
*   **Coefficient for 'Tenure_Months':** `-0.0352`
*   **Coefficient for 'Support_Calls':** `0.3825`

We also calculated the **Odds Ratios** (exponentiated coefficients):

*   **Odds Ratio for 'Monthly_Charges':** `1.0474`
*   **Odds Ratio for 'Tenure_Months':** `0.9654`
*   **Odds Ratio for 'Support_Calls':** `1.4659`

Let's interpret these values in the context of predicting customer churn:

1.  **Intercept (`-1.5969`):**
    *   The intercept represents the log-odds of a customer churning when all independent variables (Monthly Charges, Tenure Months, Support Calls) are zero. In a practical sense, a customer with zero monthly charges, zero tenure, and zero support calls is highly theoretical or non-existent. However, mathematically, this value serves as the baseline log-odds for the prediction.
    *   To convert this log-odds to a probability, we can use the sigmoid function: `P = 1 / (1 + e^(-intercept)) = 1 / (1 + e^(1.5969))`. This baseline probability is generally not directly interpretable as a 'real' probability for an average customer, but it's a foundational component of the model's structure.

2.  **Coefficient for 'Monthly_Charges' (`0.0463`):**
    *   **Log-Odds Interpretation:** For every one-unit increase in 'Monthly_Charges', the log-odds of a customer churning increase by `0.0463`, assuming all other variables are held constant.
    *   **Odds Ratio Interpretation (Odds Ratio: `1.0474`):** This means that for every $1 increase in monthly charges, the odds of a customer churning multiply by approximately `1.0474`. In other words, there's about a 4.74% increase in the odds of churn for each additional dollar in monthly charges, holding other factors constant. This suggests that higher monthly charges are associated with a higher likelihood of churn.

3.  **Coefficient for 'Tenure_Months' (`-0.0352`):**
    *   **Log-Odds Interpretation:** For every one-month increase in 'Tenure_Months', the log-odds of a customer churning decrease by `0.0352`, assuming all other variables are held constant.
    *   **Odds Ratio Interpretation (Odds Ratio: `0.9654`):** This indicates that for every additional month a customer stays, the odds of them churning multiply by approximately `0.9654`. Equivalently, there's about a `1 - 0.9654 = 0.0346` or 3.46% decrease in the odds of churn for each additional month of tenure. This aligns with intuition: longer-term customers are less likely to churn.

4.  **Coefficient for 'Support_Calls' (`0.3825`):**
    *   **Log-Odds Interpretation:** For every one-unit increase in 'Support_Calls' (i.e., one more support call), the log-odds of a customer churning increase by `0.3825`, assuming all other variables are held constant.
    *   **Odds Ratio Interpretation (Odds Ratio: `1.4659`):** This implies that for every additional support call a customer makes, the odds of them churning multiply by approximately `1.4659`. This translates to about a 46.59% increase in the odds of churn for each extra support call. This could suggest that customers who frequently contact support might be experiencing issues that lead them to churn, or are generally more dissatisfied.

**Summary of Effects:**
*   **Higher Monthly Charges** and **More Support Calls** are associated with an **increased likelihood of churn**.
*   **Longer Tenure** is associated with a **decreased likelihood of churn**.

These interpretations provide valuable insights into which factors are driving customer churn and in what direction, allowing businesses to formulate targeted strategies (e.g., improve customer service to reduce support call frequency, offer incentives to long-term customers, review pricing structures).

### Interpretation of Learned Coefficients

For our trained Logistic Regression model, the intercept and coefficients are as follows:

*   **Model Intercept (Log-Odds):** `-1.5969`
*   **Coefficient for 'Monthly_Charges':** `0.0463`
*   **Coefficient for 'Tenure_Months':** `-0.0352`
*   **Coefficient for 'Support_Calls':** `0.3825`

We also calculated the **Odds Ratios** (exponentiated coefficients):

*   **Odds Ratio for 'Monthly_Charges':** `1.0474`
*   **Odds Ratio for 'Tenure_Months':** `0.9654`
*   **Odds Ratio for 'Support_Calls':** `1.4659`

Let's interpret these values in the context of predicting customer churn:

1.  **Intercept (`-1.5969`):**
    *   The intercept represents the log-odds of a customer churning when all independent variables (Monthly Charges, Tenure Months, Support Calls) are zero. In a practical sense, a customer with zero monthly charges, zero tenure, and zero support calls is highly theoretical or non-existent. However, mathematically, this value serves as the baseline log-odds for the prediction.
    *   To convert this log-odds to a probability, we can use the sigmoid function: `P = 1 / (1 + e^(-intercept)) = 1 / (1 + e^(1.5969))`. This baseline probability is generally not directly interpretable as a 'real' probability for an average customer, but it's a foundational component of the model's structure.

2.  **Coefficient for 'Monthly_Charges' (`0.0463`):**
    *   **Log-Odds Interpretation:** For every one-unit increase in 'Monthly_Charges', the log-odds of a customer churning increase by `0.0463`, assuming all other variables are held constant.
    *   **Odds Ratio Interpretation (Odds Ratio: `1.0474`):** This means that for every $1 increase in monthly charges, the odds of a customer churning multiply by approximately `1.0474`. In other words, there's about a 4.74% increase in the odds of churn for each additional dollar in monthly charges, holding other factors constant. This suggests that higher monthly charges are associated with a higher likelihood of churn.

3.  **Coefficient for 'Tenure_Months' (`-0.0352`):**
    *   **Log-Odds Interpretation:** For every one-month increase in 'Tenure_Months', the log-odds of a customer churning decrease by `0.0352`, assuming all other variables are held constant.
    *   **Odds Ratio Interpretation (Odds Ratio: `0.9654`):** This indicates that for every additional month a customer stays, the odds of them churning multiply by approximately `0.9654`. Equivalently, there's about a `1 - 0.9654 = 0.0346` or 3.46% decrease in the odds of churn for each additional month of tenure. This aligns with intuition: longer-term customers are less likely to churn.

4.  **Coefficient for 'Support_Calls' (`0.3825`):**
    *   **Log-Odds Interpretation:** For every one-unit increase in 'Support_Calls' (i.e., one more support call), the log-odds of a customer churning increase by `0.3825`, assuming all other variables are held constant.
    *   **Odds Ratio Interpretation (Odds Ratio: `1.4659`):** This implies that for every additional support call a customer makes, the odds of them churning multiply by approximately `1.4659`. This translates to about a 46.59% increase in the odds of churn for each extra support call. This could suggest that customers who frequently contact support might be experiencing issues that lead them to churn, or are generally more dissatisfied.

**Summary of Effects:**
*   **Higher Monthly Charges** and **More Support Calls** are associated with an **increased likelihood of churn**.
*   **Longer Tenure** is associated with a **decreased likelihood of churn**.

These interpretations provide valuable insights into which factors are driving customer churn and in what direction, allowing businesses to formulate targeted strategies (e.g., improve customer service to reduce support call frequency, offer incentives to long-term customers, review pricing structures).

### Explanation of Evaluation Metrics for Classification

After training a classification model like Logistic Regression, it's crucial to evaluate its performance using appropriate metrics. Unlike regression, where errors are continuous, classification models deal with predicting discrete categories, and thus require different evaluation approaches. Here, we'll explain some of the most common classification metrics:

1.  **Accuracy**
    *   **Definition:** Accuracy measures the proportion of correctly classified instances (both true positives and true negatives) out of the total number of instances. It is the most intuitive classification metric.
    *   **Formula:** `Accuracy = (True Positives + True Negatives) / Total Samples`
    *   **Interpretation:** A higher accuracy score indicates a better performing model. However, it can be misleading in cases of imbalanced datasets.

2.  **Confusion Matrix**
    *   **Definition:** A confusion matrix is a table that is often used to describe the performance of a classification model on a set of test data for which the true values are known. It allows visualization of the performance of an algorithm.
    *   **Components:**
        *   **True Positives (TP):** Correctly predicted positive instances.
        *   **True Negatives (TN):** Correctly predicted negative instances.
        *   **False Positives (FP):** Incorrectly predicted positive instances (Type I error).
        *   **False Negatives (FN):** Incorrectly predicted negative instances (Type II error).
    *   **Interpretation:** Provides a detailed breakdown of correct and incorrect predictions for each class, helping to understand where the model is succeeding and failing.

3.  **Precision**
    *   **Definition:** Precision (also called Positive Predictive Value) measures the proportion of true positive predictions among all positive predictions made by the model. It answers the question: "Of all instances predicted as positive, how many are actually positive?"
    *   **Formula:** `Precision = True Positives / (True Positives + False Positives)`
    *   **Interpretation:** High precision means a low false positive rate. It is important when the cost of a false positive is high (e.g., flagging a non-spam email as spam).

4.  **Recall**
    *   **Definition:** Recall (also called Sensitivity or True Positive Rate) measures the proportion of true positive predictions among all actual positive instances. It answers the question: "Of all actual positive instances, how many did the model correctly identify?"
    *   **Formula:** `Recall = True Positives / (True Positives + False Negatives)`
    *   **Interpretation:** High recall means a low false negative rate. It is important when the cost of a false negative is high (e.g., failing to detect a diseased patient).

5.  **F1-Score**
    *   **Definition:** The F1-Score is the harmonic mean of Precision and Recall. It provides a single metric that balances both precision and recall, especially useful when there is an uneven class distribution.
    *   **Formula:** `F1-Score = 2 * (Precision * Recall) / (Precision + Recall)`
    *   **Interpretation:** A higher F1-Score indicates a better balance between precision and recall. It's a good measure for overall model accuracy, particularly for imbalanced datasets.

6.  **ROC AUC (Receiver Operating Characteristic - Area Under the Curve)**
    *   **Definition:** The ROC curve plots the True Positive Rate (Recall) against the False Positive Rate (1 - Specificity) at various threshold settings. The AUC (Area Under the Curve) measures the entire 2-D area underneath the entire ROC curve. It ranges from 0 to 1.
    *   **Interpretation:** An AUC of 1 represents a perfect classifier, while an AUC of 0.5 represents a classifier no better than random guessing. A higher AUC value indicates better separability between the positive and negative classes, meaning the model is better at distinguishing between them across all possible classification thresholds.

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix

# Make predictions on the test set
y_pred = model.predict(X_test)

# Get probability estimates for the positive class (class 1) for ROC AUC
y_pred_proba = model.predict_proba(X_test)[:, 1]

# Calculate Accuracy
accuracy = accuracy_score(y_test, y_pred)

# Calculate Precision
precision = precision_score(y_test, y_pred)

# Calculate Recall
recall = recall_score(y_test, y_pred)

# Calculate F1-Score
f1 = f1_score(y_test, y_pred)

# Calculate ROC AUC
roc_auc = roc_auc_score(y_test, y_pred_proba)

# Calculate Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)

print("\n--- Model Evaluation on Test Set ---")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")
print(f"ROC AUC: {roc_auc:.4f}")
print("\nConfusion Matrix:")
print(conf_matrix)


--- Model Evaluation on Test Set ---
Accuracy: 0.8050
Precision: 0.8211
Recall: 0.9689
F1-Score: 0.8889
ROC AUC: 0.8111

Confusion Matrix:
[[  5  34]
 [  5 156]]


## Interpretation of Evaluation Results

Let's interpret the calculated evaluation metrics for our Logistic Regression model on the customer churn prediction test set:

*   **Accuracy:** `0.8050`
*   **Precision:** `0.8211`
*   **Recall:** `0.9689`
*   **F1-Score:** `0.8889`
*   **ROC AUC:** `0.8111`
*   **Confusion Matrix:**
    ```
    [[  5  34]
     [  5 156]]
    ```

First, let's break down the Confusion Matrix:

*   **True Negatives (TN): `5`**
    *   These are the `5` customers who did **not churn** (actual negative) and our model correctly predicted that they would **not churn**. These are correctly identified non-churners.
*   **False Positives (FP): `34`**
    *   These are the `34` customers who did **not churn** (actual negative), but our model incorrectly predicted that they **would churn**. This is a Type I error. In a business context, these are customers we might wrongly target with retention efforts.
*   **False Negatives (FN): `5`**
    *   These are the `5` customers who **did churn** (actual positive), but our model incorrectly predicted that they **would not churn**. This is a Type II error. These are churners we failed to identify and thus couldn't intervene to retain.
*   **True Positives (TP): `156`**
    *   These are the `156` customers who **did churn** (actual positive) and our model correctly predicted that they **would churn**. These are correctly identified churners.

Now, let's interpret the other metrics in the context of customer churn:

1.  **Accuracy (`0.8050`):**
    *   **Significance:** This means our model correctly predicted the churn status for approximately `80.50%` of the customers in the test set. While `80.50%` seems good, accuracy alone can be misleading, especially if the dataset has an imbalanced class distribution (e.g., many more non-churners than churners, or vice-versa). We need to look at other metrics for a complete picture.

2.  **Precision (`0.8211`):**
    *   **Significance:** Out of all the customers our model *predicted* would churn (TP + FP = 156 + 34 = 190), `82.11%` actually did churn (`156 / 190`). A high precision means that when our model flags a customer as likely to churn, there's a high probability that they will indeed churn. This is important if the cost of incorrect churn intervention (false positive) is high.

3.  **Recall (`0.9689`):**
    *   **Significance:** Out of all the customers who *actually* churned (TP + FN = 156 + 5 = 161), our model correctly identified `96.89%` of them (`156 / 161`). A high recall means our model is very good at catching most of the actual churners. This is crucial if failing to identify a churner (false negative) is very costly (e.g., losing a valuable customer without intervention).

4.  **F1-Score (`0.8889`):**
    *   **Significance:** The F1-Score is the harmonic mean of precision and recall. Our F1-Score of `0.8889` indicates a good balance between precision and recall. This is a robust metric, especially when dealing with potential class imbalance, as it doesn't favor one type of error over another as much as precision or recall might individually.

5.  **ROC AUC (`0.8111`):**
    *   **Significance:** An ROC AUC of `0.8111` suggests that our model has a good ability to distinguish between churners and non-churners. An AUC of 1 is perfect, and 0.5 is random guessing. Our value of `0.8111` means that if we randomly pick one churner and one non-churner, our model will assign a higher probability of churning to the actual churner about 81.11% of the time. This metric is useful because it evaluates the model's performance across all possible classification thresholds.

### Overall Assessment of Model Performance

Considering all the metrics, the Logistic Regression model performs **quite well** in predicting customer churn for this synthetic dataset:

*   The **high Recall (`0.9689`)** is a significant strength. It indicates that the model successfully identifies almost all customers who are going to churn, which is often a primary business objective for churn prediction (to allow for proactive retention efforts).
*   The **good Precision (`0.8211`)** means that while the model identifies many churners, a reasonable proportion of its churn predictions are accurate, minimizing wasted retention efforts on customers who wouldn't churn anyway.
*   The **F1-Score (`0.8889`)** confirms a strong overall performance, indicating a solid balance between identifying churners and doing so accurately.
*   The **ROC AUC (`0.8111`)** further supports that the model is effective at differentiating between the two classes.

However, it's worth noting the `34` **False Positives** from the confusion matrix. These are customers predicted to churn but who actually don't. Depending on the cost of retention efforts, this number might be a point for further optimization. Conversely, the `5` **False Negatives** mean we missed predicting churn for 5 customers who eventually did churn; reducing these would also be beneficial, although the current number is very low.

In conclusion, the model demonstrates strong predictive power for identifying customer churn based on the given features. Future improvements might focus on reducing false positives or investigating the characteristics of the few false negatives to make the model even more robust.

# Task
I will generate a synthetic car price dataset with at least 1000 samples, including 'Mileage', 'Age', 'Engine Size', 'Brand', and 'Price', ensuring a linear relationship with noise, and applying one-hot encoding to the 'Brand' variable.

## Generate Synthetic Car Dataset

### Subtask:
Create a synthetic dataset for car prices with at least 1000 samples, including 'Mileage', 'Age', 'Engine Size', 'Brand', and 'Price'. Ensure a linear relationship with noise, and apply one-hot encoding to the 'Brand' variable.


**Reasoning**:
The subtask requires generating a synthetic car price dataset, including numerical features and a categorical 'Brand' feature which needs one-hot encoding. This code block will cover all the steps for data generation and preprocessing as outlined in the instructions.



In [1]:
import numpy as np
import pandas as pd

# 1. Set a random seed for reproducibility
np.random.seed(42)

# 2. Define the number of samples
n_samples = 1500

# 3. Generate arrays for the numerical independent variables:
# Mileage
mileage = np.random.normal(loc=50000, scale=30000, size=n_samples)
mileage[mileage < 0] = 0 # Ensure non-negative mileage

# Age
age = np.random.randint(1, 16, size=n_samples) # Ages from 1 to 15 years

# Engine Size
engine_size = np.random.normal(loc=2.0, scale=0.5, size=n_samples)
engine_size[engine_size < 0.5] = 0.5 # Ensure engine size is at least 0.5

# 4. Generate the categorical brand variable
brands = ['Toyota', 'Honda', 'BMW', 'Mercedes', 'Ford']
brand = np.random.choice(brands, size=n_samples)

# 5. Define intercept and coefficients for numerical variables and brand impacts
intercept = 25000 # Base price
coeff_mileage = -0.15 # Price decreases with mileage
coeff_age = -1500 # Price decreases with age
coeff_engine_size = 10000 # Price increases with engine size

# Assign different base price impacts for each brand
brand_impacts = {
    'Toyota': 5000,
    'Honda': 6000,
    'Ford': 3000,
    'BMW': 15000,
    'Mercedes': 18000
}

# 6. Calculate the price for each car with noise
noise = np.random.normal(loc=0, scale=8000, size=n_samples) # Add some random noise

price = intercept + \
        (coeff_mileage * mileage) + \
        (coeff_age * age) + \
        (coeff_engine_size * engine_size) + \
        np.array([brand_impacts[b] for b in brand]) + \
        noise

price[price < 0] = 0 # Ensure prices are non-negative

# 7. Create a Pandas DataFrame `car_df`
car_df = pd.DataFrame({
    'Mileage': mileage,
    'Age': age,
    'Engine Size': engine_size,
    'Brand': brand,
    'Price': price
})

# 8. Apply one-hot encoding to the 'Brand' column
car_df = pd.get_dummies(car_df, columns=['Brand'], drop_first=True) # drop_first to avoid multicollinearity

# 9. Print the first few rows of the generated `car_df` DataFrame
print("Synthetic Car Dataset Generated (First 5 rows):")
print(car_df.head())

# 10. Explicitly state the dependent variable and the independent variables
print(
"\n-- Variable Definitions --\n"
"Dependent Variable (Target): 'Price'\n"
"Independent Variables (Features): 'Mileage', 'Age', 'Engine Size', 'Brand_Honda', 'Brand_Mercedes', 'Brand_Toyota', 'Brand_Ford'"
)


Synthetic Car Dataset Generated (First 5 rows):
        Mileage  Age  Engine Size         Price  Brand_Ford  Brand_Honda  \
0  64901.424590   15     2.075067  28740.089674       False        False   
1  45852.070965   12     2.317865  30134.660549       False         True   
2  69430.656143   12     1.646571   9556.066071       False         True   
3  95690.895692    3     1.978648  41545.596497       False        False   
4  42975.398758    1     2.424543  43589.642846       False        False   

   Brand_Mercedes  Brand_Toyota  
0           False         False  
1           False         False  
2           False         False  
3            True         False  
4           False          True  

-- Variable Definitions --
Dependent Variable (Target): 'Price'
Independent Variables (Features): 'Mileage', 'Age', 'Engine Size', 'Brand_Honda', 'Brand_Mercedes', 'Brand_Toyota', 'Brand_Ford'


**Reasoning**:
The previous step successfully generated the synthetic car price dataset. Now, I need to implement and train a Multiple Linear Regression model using this dataset, as per the second instruction of the main task. This involves splitting the data into training and testing sets and then fitting the model.



In [2]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Define independent variables (X) and dependent variable (y)
X = car_df.drop('Price', axis=1)
y = car_df['Price']

print("X and y defined successfully.")

# Split the data into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data split into training and testing sets successfully.")
print(f"X_train shape: {X_train.shape}, y_train shape: {y_train.shape}")
print(f"X_test shape: {X_test.shape}, y_test shape: {y_test.shape}")

# Instantiate the Linear Regression model
model = LinearRegression()

# Train the model using the training data
model.fit(X_train, y_train)

print("\nLinear Regression model trained successfully.")

# Print the intercept and coefficients
print(f"\nModel Intercept: {model.intercept_:.2f}")
print("Model Coefficients:")
for feature, coef in zip(X.columns, model.coef_):
    print(f"  {feature}: {coef:.2f}")

X and y defined successfully.
Data split into training and testing sets successfully.
X_train shape: (1200, 7), y_train shape: (1200,)
X_test shape: (300, 7), y_test shape: (300,)

Linear Regression model trained successfully.

Model Intercept: 39596.64
Model Coefficients:
  Mileage: -0.16
  Age: -1540.28
  Engine Size: 10247.16
  Brand_Ford: -10988.89
  Brand_Honda: -9702.23
  Brand_Mercedes: 4377.72
  Brand_Toyota: -9023.85


### Interpretation of Learned Coefficients

The trained Multiple Linear Regression model for car price prediction has produced the following intercept and coefficients:

*   **Model Intercept:** `39596.64`
*   **Coefficient for 'Mileage':** `-0.16`
*   **Coefficient for 'Age':** `-1540.28`
*   **Coefficient for 'Engine Size':** `10247.16`
*   **Coefficient for 'Brand_Ford':** `-10988.89`
*   **Coefficient for 'Brand_Honda':** `-9702.23`
*   **Coefficient for 'Brand_Mercedes':** `4377.72`
*   **Coefficient for 'Brand_Toyota':** `-9023.85`

Let's interpret these values in the context of car price prediction:

1.  **Intercept ($39,596.64):**
    *   The intercept represents the estimated base price of a car when all independent variables are zero. In this synthetic dataset, this would theoretically be the price of a car with 0 mileage, 0 age, 0 engine size, and belonging to the baseline brand (which is 'BMW' in this case, as it was dropped during one-hot encoding). While a car with zero for all these features is not realistic, the intercept provides a starting point for the prediction, from which the effects of other features are added or subtracted.

2.  **Coefficient for 'Mileage' (-$0.16):**
    *   This coefficient indicates that, holding all other factors (age, engine size, and brand) constant, for every one-mile increase in a car's mileage, the car's price is estimated to decrease by approximately **$0.16**. This is a logical relationship, as higher mileage generally leads to lower car values.

3.  **Coefficient for 'Age' (-$1,540.28):**
    *   This coefficient suggests that, holding mileage, engine size, and brand constant, for every one-year increase in a car's age, its price is estimated to decrease by approximately **$1,540.28**. This also aligns with real-world observations, where older cars typically depreciate in value.

4.  **Coefficient for 'Engine Size' ($10,247.16):**
    *   This coefficient implies that, holding mileage, age, and brand constant, for every one-liter increase in engine size, the car's price is estimated to increase by approximately **$10,247.16**. Larger engine sizes can often correlate with more powerful or luxury vehicles, commanding a higher price.

5.  **Coefficients for 'Brand' (One-Hot Encoded):**
    *   When using `drop_first=True` for one-hot encoding, one category becomes the reference (baseline) category. In this case, 'BMW' is the baseline brand. The coefficients for the other brands represent the estimated difference in price compared to a 'BMW', holding all other numerical features constant.
        *   **Brand_Ford: (-$10,988.89):** A Ford car is estimated to be approximately **$10,988.89 cheaper** than a BMW, assuming all other car attributes are the same.
        *   **Brand_Honda: (-$9,702.23):** A Honda car is estimated to be approximately **$9,702.23 cheaper** than a BMW, assuming all other car attributes are the same.
        *   **Brand_Mercedes: ($4,377.72):** A Mercedes car is estimated to be approximately **$4,377.72 more expensive** than a BMW, assuming all other car attributes are the same. (This might seem counter-intuitive if you expect Mercedes to be much more expensive, but it reflects the synthetic data's underlying coefficients.)
        *   **Brand_Toyota: (-$9,023.85):** A Toyota car is estimated to be approximately **$9,023.85 cheaper** than a BMW, assuming all other car attributes are the same.

These interpretations provide valuable insights into how each feature contributes to the predicted car price, allowing for an understanding of the model's logic and the factors influencing car valuations within this synthetic dataset.

### Interpretation of Learned Coefficients

The trained Multiple Linear Regression model for car price prediction has produced the following intercept and coefficients:

*   **Model Intercept:** `39596.64`
*   **Coefficient for 'Mileage':** `-0.16`
*   **Coefficient for 'Age':** `-1540.28`
*   **Coefficient for 'Engine Size':** `10247.16`
*   **Coefficient for 'Brand_Ford':** `-10988.89`
*   **Coefficient for 'Brand_Honda':** `-9702.23`
*   **Coefficient for 'Brand_Mercedes':** `4377.72`
*   **Coefficient for 'Brand_Toyota':** `-9023.85`

Let's interpret these values in the context of car price prediction:

1.  **Intercept ($39,596.64):**
    *   The intercept represents the estimated base price of a car when all independent variables are zero. In this synthetic dataset, this would theoretically be the price of a car with 0 mileage, 0 age, 0 engine size, and belonging to the baseline brand (which is 'BMW' in this case, as it was dropped during one-hot encoding). While a car with zero for all these features is not realistic, the intercept provides a starting point for the prediction, from which the effects of other features are added or subtracted.

2.  **Coefficient for 'Mileage' (-$0.16):**
    *   This coefficient indicates that, holding all other factors (age, engine size, and brand) constant, for every one-mile increase in a car's mileage, the car's price is estimated to decrease by approximately **$0.16**. This is a logical relationship, as higher mileage generally leads to lower car values.

3.  **Coefficient for 'Age' (-$1,540.28):**
    *   This coefficient suggests that, holding mileage, engine size, and brand constant, for every one-year increase in a car's age, its price is estimated to decrease by approximately **$1,540.28**. This also aligns with real-world observations, where older cars typically depreciate in value.

4.  **Coefficient for 'Engine Size' ($10,247.16):**
    *   This coefficient implies that, holding mileage, age, and brand constant, for every one-liter increase in engine size, the car's price is estimated to increase by approximately **$10,247.16**. Larger engine sizes can often correlate with more powerful or luxury vehicles, commanding a higher price.

5.  **Coefficients for 'Brand' (One-Hot Encoded):**
    *   When using `drop_first=True` for one-hot encoding, one category becomes the reference (baseline) category. In this case, 'BMW' is the baseline brand. The coefficients for the other brands represent the estimated difference in price compared to a 'BMW', holding all other numerical features constant.
        *   **Brand_Ford: (-$10,988.89):** A Ford car is estimated to be approximately **$10,988.89 cheaper** than a BMW, assuming all other car attributes are the same.
        *   **Brand_Honda: (-$9,702.23):** A Honda car is estimated to be approximately **$9,702.23 cheaper** than a BMW, assuming all other car attributes are the same.
        *   **Brand_Mercedes: ($4,377.72):** A Mercedes car is estimated to be approximately **$4,377.72 more expensive** than a BMW, assuming all other car attributes are the same. (This might seem counter-intuitive if you expect Mercedes to be much more expensive, but it reflects the synthetic data's underlying coefficients.)
        *   **Brand_Toyota: (-$9,023.85):** A Toyota car is estimated to be approximately **$9,023.85 cheaper** than a BMW, assuming all other car attributes are the same.

These interpretations provide valuable insights into how each feature contributes to the predicted car price, allowing for an understanding of the model's logic and the factors influencing car valuations within this synthetic dataset.

**Reasoning**:
The previous step interpreted the model coefficients. Now, as per the main task's instructions (point 4), I need to evaluate the model's performance on the test set by calculating and printing R-squared, MAE, MSE, and RMSE.



In [3]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate R-squared
r2 = r2_score(y_test, y_pred)

# Calculate Mean Absolute Error (MAE)
mae = mean_absolute_error(y_test, y_pred)

# Calculate Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)

# Calculate Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)

print("\n--- Model Evaluation on Test Set ---")
print(f"R-squared: {r2:.4f}")
print(f"Mean Absolute Error (MAE): ${mae:,.2f}")
print(f"Mean Squared Error (MSE): ${mse:,.2f}")
print(f"Root Mean Squared Error (RMSE): ${rmse:,.2f}")


--- Model Evaluation on Test Set ---
R-squared: 0.6256
Mean Absolute Error (MAE): $6,402.62
Mean Squared Error (MSE): $68,169,437.76
Root Mean Squared Error (RMSE): $8,256.48


### Interpretation of Evaluation Results

Let's review the calculated evaluation metrics for our Multiple Linear Regression model on the car price prediction test set:

*   **R-squared:** `0.6256`
*   **Mean Absolute Error (MAE):** `$6,402.62`
*   **Mean Squared Error (MSE):** `$68,169,437.76`
*   **Root Mean Squared Error (RMSE):** `$8,256.48`

#### 1. R-squared (Coefficient of Determination)
*   **Significance:** R-squared measures the proportion of the variance in the dependent variable (Car Price) that can be explained by the independent variables (Mileage, Age, Engine Size, and Brand) in our model. It ranges from 0 to 1, where higher values indicate a better fit.
*   **Our Result (0.6256):** An R-squared value of `0.6256` (or approximately 62.56%) means that about 62.56% of the variability in car prices can be explained by our model's chosen features. The remaining 37.44% of the variability is still unexplained, likely due to other factors not included in our dataset or inherent randomness. While not extremely high, it suggests a moderate explanatory power for this synthetic dataset.

#### 2. Mean Absolute Error (MAE)
*   **Significance:** MAE measures the average magnitude of the errors in a set of predictions, without considering their direction. It is expressed in the same units as the dependent variable, making it easily interpretable.
*   **Our Result ($6,402.62):** This means that, on average, our model's predictions for car prices are off by approximately **$6,402.62** from the actual prices. This provides a direct understanding of the typical prediction error.

#### 3. Mean Squared Error (MSE)
*   **Significance:** MSE measures the average of the squares of the errors, penalizing larger errors more heavily than smaller ones. It is in squared units of the dependent variable.
*   **Our Result ($68,169,437.76):** This value is less intuitive for direct interpretation in dollar terms due to being in squared dollars. It is more useful for mathematical optimization during model training.

#### 4. Root Mean Squared Error (RMSE)
*   **Significance:** RMSE is the square root of MSE, bringing the error metric back into the same units as the dependent variable. It represents the standard deviation of the residuals (prediction errors) and gives higher weight to larger errors.
*   **Our Result ($8,256.48):** The RMSE of **$8,256.48** tells us that the typical difference between our predicted car prices and the actual car prices is around this amount. The fact that RMSE is slightly higher than MAE suggests that there might be some larger individual prediction errors that are being more heavily penalized by RMSE.

#### Overall Assessment of Model Performance

Considering these metrics, the Multiple Linear Regression model performs reasonably for this synthetic car price dataset:

*   The **R-squared of 0.6256** indicates that the model captures a significant portion of the car price variance, although there is still substantial unexplained variability.
*   The **MAE of $6,402.62** and **RMSE of $8,256.48** provide a concrete measure of the prediction accuracy. For car prices, which can range widely, an average error of around $6,000-$8,000 might be considered acceptable depending on the specific application and the typical range of prices in the dataset. This suggests that the model provides fairly good estimates.

In a real-world scenario, the acceptability of these error margins would depend on the business context (e.g., whether these errors are within typical profit margins or acceptable bidding ranges for a dealership). Given that this is a synthetic dataset with introduced noise, the results demonstrate that the model successfully learned the underlying linear relationships.

### Interpretation of Evaluation Results

Let's review the calculated evaluation metrics for our Multiple Linear Regression model on the car price prediction test set:

*   **R-squared:** `0.6256`
*   **Mean Absolute Error (MAE):** `$6,402.62`
*   **Mean Squared Error (MSE):** `$68,169,437.76`
*   **Root Mean Squared Error (RMSE):** `$8,256.48`

#### 1. R-squared (Coefficient of Determination)
*   **Significance:** R-squared measures the proportion of the variance in the dependent variable (Car Price) that can be explained by the independent variables (Mileage, Age, Engine Size, and Brand) in our model. It ranges from 0 to 1, where higher values indicate a better fit.
*   **Our Result (0.6256):** An R-squared value of `0.6256` (or approximately 62.56%) means that about 62.56% of the variability in car prices can be explained by our model's chosen features. The remaining 37.44% of the variability is still unexplained, likely due to other factors not included in our dataset or inherent randomness. While not extremely high, it suggests a moderate explanatory power for this synthetic dataset.

#### 2. Mean Absolute Error (MAE)
*   **Significance:** MAE measures the average magnitude of the errors in a set of predictions, without considering their direction. It is expressed in the same units as the dependent variable, making it easily interpretable.
*   **Our Result ($6,402.62):** This means that, on average, our model's predictions for car prices are off by approximately **$6,402.62** from the actual prices. This provides a direct understanding of the typical prediction error.

#### 3. Mean Squared Error (MSE)
*   **Significance:** MSE measures the average of the squares of the errors, penalizing larger errors more heavily than smaller ones. It is in squared units of the dependent variable.
*   **Our Result ($68,169,437.76):** This value is less intuitive for direct interpretation in dollar terms due to being in squared dollars. It is more useful for mathematical optimization during model training.

#### 4. Root Mean Squared Error (RMSE)
*   **Significance:** RMSE is the square root of MSE, bringing the error metric back into the same units as the dependent variable. It represents the standard deviation of the residuals (prediction errors) and gives higher weight to larger errors.
*   **Our Result ($8,256.48):** The RMSE of **$8,256.48** tells us that the typical difference between our predicted car prices and the actual car prices is around this amount. The fact that RMSE is slightly higher than MAE suggests that there might be some larger individual prediction errors that are being more heavily penalized by RMSE.

#### Overall Assessment of Model Performance

Considering these metrics, the Multiple Linear Regression model performs reasonably for this synthetic car price dataset:

*   The **R-squared of 0.6256** indicates that the model captures a significant portion of the car price variance, although there is still substantial unexplained variability.
*   The **MAE of $6,402.62** and **RMSE of $8,256.48** provide a concrete measure of the prediction accuracy. For car prices, which can range widely, an average error of around $6,000-$8,000 might be considered acceptable depending on the specific application and the typical range of prices in the dataset. This suggests that the model provides fairly good estimates.

In a real-world scenario, the acceptability of these error margins would depend on the business context (e.g., whether these errors are within typical profit margins or acceptable bidding ranges for a dealership). Given that this is a synthetic dataset with introduced noise, the results demonstrate that the model successfully learned the underlying linear relationships.

### Interpretation of Evaluation Results

Let's review the calculated evaluation metrics for our Multiple Linear Regression model on the car price prediction test set:

*   **R-squared:** `0.6256`
*   **Mean Absolute Error (MAE):** `$6,402.62`
*   **Mean Squared Error (MSE):** `$68,169,437.76`
*   **Root Mean Squared Error (RMSE):** `$8,256.48`

#### 1. R-squared (Coefficient of Determination)
*   **Significance:** R-squared measures the proportion of the variance in the dependent variable (Car Price) that can be explained by the independent variables (Mileage, Age, Engine Size, and Brand) in our model. It ranges from 0 to 1, where higher values indicate a better fit.
*   **Our Result (0.6256):** An R-squared value of `0.6256` (or approximately 62.56%) means that about 62.56% of the variability in car prices can be explained by our model's chosen features. The remaining 37.44% of the variability is still unexplained, likely due to other factors not included in our dataset or inherent randomness. While not extremely high, it suggests a moderate explanatory power for this synthetic dataset.

#### 2. Mean Absolute Error (MAE)
*   **Significance:** MAE measures the average magnitude of the errors in a set of predictions, without considering their direction. It is expressed in the same units as the dependent variable, making it easily interpretable.
*   **Our Result ($6,402.62):** This means that, on average, our model's predictions for car prices are off by approximately **$6,402.62** from the actual prices. This provides a direct understanding of the typical prediction error.

#### 3. Mean Squared Error (MSE)
*   **Significance:** MSE measures the average of the squares of the errors, penalizing larger errors more heavily than smaller ones. It is in squared units of the dependent variable.
*   **Our Result ($68,169,437.76):** This value is less intuitive for direct interpretation in dollar terms due to being in squared dollars. It is more useful for mathematical optimization during model training.

#### 4. Root Mean Squared Error (RMSE)
*   **Significance:** RMSE is the square root of MSE, bringing the error metric back into the same units as the dependent variable. It represents the standard deviation of the residuals (prediction errors) and gives higher weight to larger errors.
*   **Our Result ($8,256.48):** The RMSE of **$8,256.48** tells us that the typical difference between our predicted car prices and the actual car prices is around this amount. The fact that RMSE is slightly higher than MAE suggests that there might be some larger individual prediction errors that are being more heavily penalized by RMSE.

#### Overall Assessment of Model Performance

Considering these metrics, the Multiple Linear Regression model performs reasonably for this synthetic car price dataset:

*   The **R-squared of 0.6256** indicates that the model captures a significant portion of the car price variance, although there is still substantial unexplained variability.
*   The **MAE of $6,402.62** and **RMSE of $8,256.48** provide a concrete measure of the prediction accuracy. For car prices, which can range widely, an average error of around $6,000-$8,000 might be considered acceptable depending on the specific application and the typical range of prices in the dataset. This suggests that the model provides fairly good estimates.

In a real-world scenario, the acceptability of these error margins would depend on the business context (e.g., whether these errors are within typical profit margins or acceptable bidding ranges for a dealership). Given that this is a synthetic dataset with introduced noise, the results demonstrate that the model successfully learned the underlying linear relationships.


## Final Task

### Subtask:
Summarize the findings from the car price prediction model, including key insights from coefficient interpretation and overall model performance.


## Summary:

### Q&A
The car price prediction model reveals that several factors significantly influence car prices. Mileage and age negatively impact price, while engine size positively affects it. Different car brands also have varying price effects, with Mercedes-Benz cars being more expensive than BMWs, and Ford, Honda, and Toyota being less expensive than BMWs, all else being equal. The model explains a moderate amount of price variability and has an average prediction error in the range of \\$6,000 to \\$8,000.

### Data Analysis Key Findings
*   **Model Performance**: The Linear Regression model achieved an R-squared value of 0.6256, indicating that approximately 62.56% of the variability in car prices can be explained by the included features.
*   **Prediction Accuracy**: The model exhibited a Mean Absolute Error (MAE) of \\$6,402.62, meaning that, on average, its predictions deviate from actual prices by this amount. The Root Mean Squared Error (RMSE) was \\$8,256.48.
*   **Impact of Mileage**: For every additional mile, the car's price is estimated to decrease by approximately \\$0.16, holding other factors constant.
*   **Impact of Age**: Each additional year of a car's age is associated with an estimated price decrease of \\$1,540.28, assuming other factors remain constant.
*   **Impact of Engine Size**: An increase of one liter in engine size is estimated to increase the car's price by \\$10,247.16, all else being equal.
*   **Brand-Specific Price Adjustments (relative to BMW)**:
    *   A Ford car is estimated to be \\$10,988.89 cheaper than a BMW.
    *   A Honda car is estimated to be \\$9,702.23 cheaper than a BMW.
    *   A Mercedes car is estimated to be \\$4,377.72 more expensive than a BMW.
    *   A Toyota car is estimated to be \\$9,023.85 cheaper than a BMW.
*   **Baseline Price**: The model's intercept is \\$39,596.64, representing the theoretical base price for a car with zero mileage, age, engine size, and belonging to the baseline 'BMW' brand.

### Insights or Next Steps
*   The model successfully captured the predefined linear relationships in the synthetic dataset, demonstrating that mileage and age depreciate car value, while engine size and certain brands contribute to higher prices.
*   To improve model performance and explanatory power, especially given the R-squared of 0.6256, future steps could involve introducing additional features such as car condition, transmission type, trim level, or incorporating non-linear relationships if applicable to a real-world scenario.
